File: Rust/src/donp/transform/conversion.rs
/*
	Package 'transform' provides mechanism's for converting data between different formats.
*/
/*
	The Conversion module provides methods to convert data between hexadecimal and ASCII formats,
	according to the configured transmission mode.
*/

use log::error;

pub const HEXADECIMAL_MODE: &str = "hex";
const ASCII_MODE: &str = "ascii";

pub struct Conversion{
    mode: String,    
}

impl Default for Conversion {
    fn default() -> Self {
        Conversion {
            mode: String::new(),
        }
    }
}

impl Conversion {
    pub fn new(mode: &str) -> Conversion {
        Conversion {
            mode: mode.to_string(),
        }
    }

    pub fn clone(&self) -> Conversion {
        Conversion {
            mode: self.mode.clone(),
        }
    }

    pub fn get_mode(&self) -> &String {
        &self.mode
    }

    pub fn from_hex_to_ascii(&self, message: &Vec<u8>, prefix: &str, suffix: &str) -> Vec<u8> {
        // Convert a hexadecimal byte array to its ASCII representation.
        let mut decoded = Vec::<u8>::new();
        let length = message.len();
        let mut index = 0;
        let mut byte_string: String;
        for byte in message {
            byte_string = format!("{:02X}", byte);
            let low = byte_string[0..1].chars().nth(0).unwrap_or_default() as char;
            let high = byte_string[1..2].chars().nth(0).unwrap_or_default() as char;
            if self.is_ascii(low as u8) && self.is_ascii(high as u8) {
                decoded.extend(format!("{}{}", low as char, high as char).as_bytes().to_vec());
            } else if self.is_allowed(*byte, prefix, suffix, index as i32, length as i32) {
                decoded.push(*byte);
            }
            index += 1;           
        }
        decoded
    }
                
    
    pub fn from_ascii_to_hex(&self, message: &Vec<u8>, prefix: &str, suffix: &str) -> Vec<u8> {
        // Convert an ASCII byte array to its hexadecimal representation.
        let mut encoded = Vec::<u8>::new();
        let length = message.len();
        let mut index = 0;
        while index < length {
            if self.is_allowed(message[index], prefix, suffix, index as i32, length as i32) {
                encoded.push(message[index]);
                index += 1;
                continue;
            }
            if self.is_ascii(message[index]) {                
                if index + 1 < length && self.is_ascii(message[index+1]){
                    let hex_string = String::from_utf8(vec![message[index], message[index+1]]).unwrap_or("00".to_string());
                    let hex_byte = u8::from_str_radix(&hex_string, 16).unwrap_or(0);
                    encoded.push(hex_byte);
                    index += 2;
                    continue;
                }
            }
            error!("Invalid ASCII character in message: {}", message[index] as char);
            index += 1;
        }
        encoded
    }

    pub fn is_ascii(&self, byte: u8) -> bool {
        // Check if a byte represents a valid ASCII hexadecimal character.
        (byte >= b'0' && byte <= b'9') ||
        (byte >= b'A' && byte <= b'F') ||
        (byte >= b'a' && byte <= b'f')
    }

    pub fn is_allowed(&self, b: u8, prefix: &str, suffix: &str, index: i32, length: i32) -> bool {
        // Check if a byte is allowed as part of the prefix or suffix.
        if b <= 127 {            
            if prefix.contains(b as char) && index < prefix.len() as i32 {
                return true
            } 
            if suffix.contains(b as char) && index >= (length - suffix.len() as i32) {
                return true
            }
        }        
        return false
    }

    pub fn display(&self, message: &Vec<u8>, prefix: &str, suffix: &str) -> String {
        // Generate a string representation of the message for display purposes.
        let mut display: Vec<u8> = message.clone();
        if self.mode == ASCII_MODE {
            display = self.from_ascii_to_hex(message, prefix, suffix)
        }
        let mut display_message = String::new();
        for byte in display {
            display_message.push_str(&format!("{:02X}", byte));
        }
        display_message.trim().to_string()
    }

    pub fn get_hex_message(&self, message: &Vec<u8>, prefix: &str, suffix: &str) -> Vec<u8> {
        // Retrieves the hexadecimal message based on the conversion mode.
        if self.mode == ASCII_MODE {
            return self.from_ascii_to_hex(message, prefix, suffix);
        }
        message.clone()
    }

    pub fn get_converted_message(&self, message: &Vec<u8>, prefix: &str, suffix: &str) -> Vec<u8> {
        // Retrieves the message based on the conversion mode.
        if self.mode == ASCII_MODE {
            return self.from_hex_to_ascii(message, prefix, suffix);
        }    
        message.clone()
    }
}

File: C++/pkg/transform/conversion.cpp
/*
	Package 'transform' provides mechanism's for converting data between different formats.
*/
/*
	The Conversion module provides methods to convert data between hexadecimal and ASCII formats,
	according to the configured transmission mode.
*/

#include "pkg/transform/conversion.hpp"

// Constructor
Conversion::Conversion() : mode(HEXADECIMAL_MODE) {}

// Destructor
Conversion::~Conversion() {
    // No dynamic memory to free
}

// Methods
std::string Conversion::get_mode() {
    return mode;
}

std::string Conversion::set_mode(std::string new_mode) {
    if (new_mode == HEXADECIMAL_MODE || new_mode == ASCII_MODE) {
        mode = new_mode;
    }
    return mode;
}

VecU8 Conversion::from_hex_to_ascii(const VecU8& hex_data, std::string prefix, std::string suffix) {
    // Convert a hexadecimal byte array to its ASCII representation.
    VecU8 ascii = {};
    std::size_t length = hex_data.size();
    std::uint32_t index = 0;
    char byte_str[3];
    for ( auto const byte : hex_data) {
        std::snprintf(byte_str, sizeof(byte_str), "%02X", byte);
        if (is_ascii(int(byte_str[0])) && is_ascii(int(byte_str[1]))) {
            ascii.push_back(static_cast<uint8_t>(byte_str[0]));
            ascii.push_back(static_cast<uint8_t>(byte_str[1]));
        } else if (is_allowed(std::byte{static_cast<uint8_t>(byte)}, prefix, suffix, index, length)) {
            ascii.push_back(static_cast<uint8_t>(byte));
        }
        index++;    
    }
    return ascii;
}

VecU8 Conversion::from_ascii_to_hex(const VecU8& ascii_data, std::string prefix, std::string suffix) {
    // Convert an ASCII byte array to its hexadecimal representation.
    VecU8 hex = {};
    std::size_t length = ascii_data.size();
    std::uint32_t index = 0;
    while (index < length) {
        if (is_allowed(std::byte{static_cast<uint8_t>(ascii_data[index])}, prefix, suffix, index, length)) {
            hex.push_back(static_cast<uint8_t>(ascii_data[index]));
            index++;
            continue;
        }
        if (is_ascii(int(ascii_data[index]))) {
            if (index + 1 < length && is_ascii(int(ascii_data[index + 1]))) {
                std::string hex_str;
                hex_str += static_cast<char>(ascii_data[index]);
                hex_str += static_cast<char>(ascii_data[index + 1]);
                unsigned int hex_value = std::stoul(hex_str, nullptr, 16);
                hex.push_back(static_cast<uint8_t>(hex_value));
                index += 2;
                continue;
            }
       }
        // Invalid character encountered; skip it
        index++;
    }
    return hex;  
}

std::string Conversion::display(const VecU8& data, std::string prefix, std::string suffix) {
    // Generate a string representation of the message for display purposes.
    VecU8 display = data;
    if (mode == ASCII_MODE) {
        display = from_ascii_to_hex(data, prefix, suffix);
    }

    // Convert vector to hexadecimal string output
    std::string display_str;
    for (const auto& byte : display) {
        char buffer[3];
        std::snprintf(buffer, sizeof(buffer), "%02X", static_cast<unsigned int>(byte));
        display_str += buffer;
    }
    display_str += "\0";
    return display_str;
}

VecU8 Conversion::get_hex_message(const VecU8& data, std::string prefix, std::string suffix) {
    // Retrieve the hexadecimal representation of the message based on the conversion mode.
    if (mode == ASCII_MODE) {
        return from_ascii_to_hex(data, prefix, suffix);
    }
    return data;
}

VecU8 Conversion::get_converted_message(const VecU8& data, std::string prefix, std::string suffix) {
    // Retrieve the converted message based on the conversion mode.
    if (mode == ASCII_MODE) {
        return from_hex_to_ascii(data, prefix, suffix);
    } 
    return data;
}

bool Conversion::is_ascii(const int data) {
    // Check if the provided byte is a valid ASCII character.
    if (data != 0x00 && data <= 0x7F) {
        if (data >= 0x30 && data <= 0x39 || 
            data >= 0x41 && data <= 0x46 || 
            data >= 0x61 && data <= 0x66) {
            return true;
        }
    }
    return false;
}

bool Conversion::is_allowed(const std::byte data, std::string prefix, std::string suffix, int index, int length) {
    // Check if the provided byte is allowed in ASCII mode.
    if (data < std::byte{0x7F}) {
        size_t find = prefix.find(static_cast<char>(data));
        if (find < prefix.length() && index < prefix.length()) {
            return true;
        }
        find = suffix.find(static_cast<char>(data));
        if (find < suffix.length() && index >= (length - suffix.length())) {
            return true;
        }
    }
    return false;
}